### ISSUE TITLE
Automatic MTU configuration / discovery

### ISSUE THREAD

**Issue by plwhite** [projectcalico/calicoctl#488]
The veth created by calicoctl has an MTU of 1500. Unfortunately, if we are running in an environment where we have a very large MTU (such as on AWS, but also private data clouds), then we might want that MTU to be larger, and in some odd clouds with small MTUs we might want it to be smaller.

The limits are the following (breaking any of these will impact performance).
- The MTU should not be larger than the MTU of the main interface to the host.
- The MTU should not be larger than 1500 if we are concerned about traffic to / from the internet
- The MTU should not be larger than the MTU of the tunnel interface if any (for which the same statements apply).
- Given the above, the MTU should be as big as possible.

Net results is that the MTU for the veths we create should be configurable. Configuration at a global level for all new veths seems like the best idea right now, with a default based on the tunnel veth (if any), otherwise 1500. A future enhancement might allow it to be configured on a per-container basis.

**Comment by plwhite** [comment]
One minor tweak - we absolutely ought to document the "right" values to set in the AWS instructions, so people know what to do here! This may well already be in hand, but just in case.

**Comment by tomdee** [comment]
We're not planning to fix this in the near future. Please chime in if you think that position is incorrect.

**Comment by plwhite** [comment]
I think it would be good to fix this, as it's a bit embarrassing that our network performance isn't great out the box. However, I don't have any reason to push this as a high priority, so back burner is probably fine.

**Comment by TrimBiggs** [comment]
We've had a few more discussions about MTU configuration via calicoctl.  We're planning to run a couple of tests to get a better idea of how Path MTU discovery works between containers and hosts.  We'll keep this issue updated with the progress towards a solution.

**Comment by spikecurtis** [comment]
@TrimBiggs Did a test where we set up two containers, running over IPIP with jumbo-frame MTU on the real interface and ipip tunnel.  We start an iperf test, and then set the MTU on the server ipip tunnel interface back to 1500.

This results in some really bad TCP behaviour.  On the server, we see all kinds of out of order packets, enough to cause the server to start using selective ACK.  The packets are still jumbo sized, however, so the kernel must be doing something strange with them while decapsulating---maybe some kind of generic receive segmentation offload, but I'm not sure.

Moral of the story is that it is bad news to change the MTU of the tunl0 interface while TCP sessions are up.

**Comment by TrimBiggs** [comment]
I've just added a section to the AWS doc that explains how to manually set MTU for jumbo-frames: #521

**Comment by titilambert** [comment]
Hello, I got a problem with MTU. MTU of container cali interface should be less or equal to the MTU of tunl0 interface. (I think the default is 1440)
Also, It should be cool, to be able to define the MTU when we create docker interface.
```
docker network create --driver=calico --opt nat-outgoing=true --opt ipip=true --mtu=1440 --subnet=192.168.1.0/24
```

Thanks !

**Comment by TrimBiggs** [comment]
Thanks for finding this @titilambert, sorry it's caused you trouble.

I've moved your comment to be a new issue (#610) since the issue here describes an enhancement to modify the MTU.

**Comment by caseydavenport** [comment]
Not convinced this should be calicoctl's job - the next CNI plugin release, for example, will support this.

**Comment by yottatsa** [comment]
How about https://github.com/projectcalico/calico-cni/commit/48168e268a516738565cb782888958e6fb0ae20a ?

**Comment by caseydavenport** [comment]
@yottatsa Yes, the Calico CNI plugin does this now :)

Leaving this open though since not all integrations use CNI.

**Comment by hectorj2f** [comment]
@caseydavenport How could I configure the MTU of my `tunl0` ? My calico interfaces for the containers use MTU 1500 while my `tunl0` has `1440`. I know that I can modify the MTU of the calico containers with the new version. But I am not able to change the MTU of the `tunl0` for the ip-ip. How could we do that ?

I have a k8s cluster that uses calico as cni plugin.

**Comment by caseydavenport** [comment]
@hectorj2f - a little bit of digging found this: https://github.com/projectcalico/calico/blob/master/calico/felix/config.py#L279

It looks like you can configure the tunl0 interface MTU as Felix config.

**Comment by hectorj2f** [comment]
@caseydavenport how can we configure it running `calico/node` using `calicoctl` ? 
`/opt/bin/calicoctl node --ip=${DEFAULT_IPV4} --detach=false --node-image=calico/node:v0.21.0`

I saw you don't have arguments to pass the felix configuration. I understand that the only solution is to manually run calico/node without calicoctl. Am I wrong ?

**Comment by caseydavenport** [comment]
Hmm.  Yes, the easiest way is probably to run `calico/node` without using `calicoctl`.  This will let you pass in whatever environment / file based configuration you'd like to.  I would have hoped that `calicoctl config` would let you set this, but apparently it does not.

I'd like to start moving our documentation that direction more generally anyway, since it's a more flexible and understandable way of launching `calico/node`.

Additionally, I think the WIP Golang `calicoctl` will support more config options than are currently exposed through calicoctl, but there isn't a release of that feature yet.

**Comment by caseydavenport** [comment]
Linking related issue (https://github.com/projectcalico/calico-containers/issues/610) and re-purposing to cover general MTU configuration support, where I think the current status is like so:
- Calico CNI plugin allows configuration of container veth MTU, default 1500.
- Felix allows configuration of tunl0 MTU, default 1440 (but this isn't exposed through calicoctl)
- Docker libnetwork does not provide a way to configure container veth MTU (?)
- We need to make sure this is all documented.
- We might want to add some sort of auto-configuration support somewhere (separate issue)

**Comment by hectorj2f** [comment]
I'd also like to be able to configure it via `calicoctl config`. Thanks for the effort @caseydavenport. I'll keep 👀  in these issues.

**Comment by simonswine** [comment]
I would love to see better support (or documentation for custom MTU settings)

* To setup the value for the `veth`/`caliXX` devices, you have to use the field in the cni-config:
```
 "mtu": 8981,
```
* The generic `tun0` device can be raised using an environment setting for felix/calico-node container:
```
FELIX_IPINIPMTU="8981"
```

This is based on AWS' 9001 MTU

**Comment by caseydavenport** [comment]
Been a while since I've looked at this guy.

- There is now documentation covering this here: http://docs.projectcalico.org/v2.3/usage/configuration/mtu
- Folks should be able to configure Felix via calicoctl a la `calicoctl config set --raw=felix IpInIpMtu  X` (disclaimer I haven't tried it myself yet)

There's still some things we _might_ want to do:
- Ability to configure the container veth via calicoctl (add another config option, make CNI plugin respect it)
- Ability to auto detect MTU and set configuration options accordingly.

**Comment by ti-mo** [comment]
Hi @caseydavenport 

Mesos/Marathon user with IPinIP and Docker containerizer here. We're able to set the MTU on our CNI workloads, but since most of them are Docker-based (and the Unified Containerizer just isn't an option right now), we cannot benefit from this feature in CNI.

Tuning `tunl0` on the host using `calicoctl config set --raw=felix IpInIpMtu 8981` works perfectly, but the interfaces inside the containers unfortunately don't follow suit.

The nature of the problem is not clear to me, but I'm fairly uninitiated on the architecture of the Docker plugin system; is it `libnetwork` that doesn't expose an API for setting MTU on endpoints or is there just a gap in the calico plugin that requires some legwork to implement? Passing `--opt` to `docker network create` is actively disallowed, I'm very glad there was user feedback in place. :)

Clearly if this was low-hanging fruit, it would have been done ages ago, so I'm wondering what kind of effort we're looking at to get this implemented.

**Comment by caseydavenport** [comment]
@ti-mo I actually don't think adding an equivalent configuration option to the libnetwork plugin would be particularly difficult.

See here for another example of a config option: https://github.com/projectcalico/libnetwork-plugin/blob/353fc3dccbc123caf9010a3dca9ba0485143aa5f/driver/network_driver.go#L69

Adding a similar `CALICO_LIBNETWORK_MTU` option would probably be relatively straightforward, provided you just want a 1:1 equivalent of the CNI feature.

**Comment by ti-mo** [comment]
@caseydavenport Thanks so much for the quick reply, we have an internal tryout running successfully. 👍 

Which approach would you prefer, extending the signature of `CreateVeth()` to include MTU, or adding a `SetVethMTU()` method? In the `SetVethMTU()` case, both calls should be executed from `NetworkDriver.Join()`, and it's an extra netlink call. Is this something you'd accept to be merged upstream?

IMO we could as well use the `IpInIpMtu` value for this is possible, 'discovering' MTU doesn't make sense to me, as only one MTU value can be set on each container interface. This forces the user to pick a single common denominator MTU across the whole cluster; tunnels over the internet must be <1500 anyway, there's no way for the sender to know which nodes in the mesh are local and which aren't.

**Comment by caseydavenport** [comment]
@ti-mo I would have thought extending `CreateVeth` to be optimal.

> IMO we could as well use the IpInIpMtu value for this is possible

We probably could, but it's more flexible to be able to configure the container veth and IPIP tunnel interfaces separately, so let's stick with that approach for now. There are scenarios where it's desirable to have two knobs to adjust here rather than one.

**Comment by caseydavenport** [comment]
whoops this should be closed now in v3.17!

**Comment by hectorj2f** [comment]
Thanks @caseydavenport 🎉
